Tx Relay v2: New Relay Logic with threshold-based peer dropping

0xFFFC0000 opened the pull request:
## Tx Relay v2: Enhanced Transaction Propagation Protocol https://github.com/monero-project/monero/issues/9334

This was made possible thanks to @Boog900’s support in designing and implementing each step.

### Key Improvements & Technical Implementation

#### 1. Inventory-Based Transaction Announcements
**Protocol Messages:**
```cpp
NOTIFY_TX_POOL_INV (ID 2011):
- Broadcasts transaction hashes instead of full TXs
- Contains vector of crypto::hash entries

NOTIFY_REQUEST_TX_POOL_TXS (ID 2012):
- Requests specific TXs by hash
- Enables selective fetching of missing transactions
```

**Workflow**:

- Node advertises new TX hashes via `NOTIFY_TX_POOL_INV`

- Receivers check local mempool & request missing TXs with `NOTIFY_REQUEST_TX_POOL_TXS`

- Only requested transactions are transmitted

**Bandwidth Savings**:

- Reduces initial broadcast size by ~99% (hash vs full TX)

#### Peer Reliability System
Peer Scoring Metrics (`peerinfo_manager`):

- Tracked per connection:
  * Announcements received
  * Successful TX deliveries
  * Missed requests
  * Request/response ratios
  * Timestamps for temporal analysis

- Threshold-based eviction:
  if (missed_requests / total_announcements) > 70%: disconnect_peer()
    
**Failure Detection**:

- Tracks 3 consecutive failures for any TX request

- Uses sliding window (last 30 requests) for fairness

- Penalizes peers for:

  * Unresponsive behavior

  * Invalid transaction responses

#### Components:
```
[Request Manager] <-handles-> [TX Request Queue]
[TX Request Queue] <-tracks-> [Peer UUIDs]
[TX Handler] <-manages-> [Timeouts]
```

#### Configuration

```cpp
 // cryptonote_config.h
#define P2P_DEFAULT_REQUEST_TIMEOUT 5      // Seconds
#define P2P_REQUEST_FAILURE_THRESHOLD 70   // Percentage
```

SChernykh commented on src/cryptonote_basic/connection_context.cpp:
```diff
@@ -65,6 +65,10 @@ namespace cryptonote
+    case cryptonote::NOTIFY_TX_POOL_INV::ID:
+      return 1024 * 1024 * 16; // 16 MB
+    case cryptonote::NOTIFY_REQUEST_TX_POOL_TXS::ID:
+      return 1024 * 1024 * 16; // 16 MB
```
16 MB is 524288 32-byte hashes. Do we really need to request 500k transactions at a time? I think a few thousand will be max, so 1 MB will be more than enough.

nahuhh replied:
Just a note: the default txpool config is something like 648mb, or about max 432k txs

SChernykh replied:
Maybe this limit makes sense then. But it doesn't have to be this big anyway - peers can send multiple messages if there are too many transactions.

Boog900 replied:
These messages should contain no where near the number of txs as there are in the pool. I would say 1 MB would be good, if we really wanted to push it we could go with the amount of txs needed to reach the 100 MB packet limit in `NOTIFY_NEW_TRANSACTIONS`: 100_000_000 / 1_500 = 66_666, 66_666 * 32 = 2_133_333. So 2 MB.

`NOTIFY_GET_TXPOOL_COMPLEMENT` which contains the ID of every tx in our pool itself has a 4 MB limit.

0xFFFC0000 replied:
Thanks to everyone participating in this discussion. I used 16MB as placeholders, since the tx-relay-v2 parameters will, of course, be a discussion topic when reviewing. I agree that @Boog900’s suggestion is a reasonable approach. Other limitations will kick in even before approaching 16MB.

nahuhh replied:
Is the padding still there?

@boog900 thoughts on padding the hashes? 
My thought was that its unnecessary. Padding txs makes sense (privacy. Different txs have different sizes) but not the hashes, which should be(?) uniform in size

0xFFFC0000 replied:
Fixed. Feel free to resolve it.

https://github.com/monero-project/monero/compare/d024d369d590a0bd6488757266fee0c45df217c7..3fcd5c70cf964b4f283c92538c81f2f2aed84c53

SChernykh commented on src/cryptonote_config.h:
```diff
@@ -151,14 +152,16 @@
 
 #define P2P_SUPPORT_FLAG_FLUFFY_BLOCKS                  0x01
-#define P2P_SUPPORT_FLAGS                               P2P_SUPPORT_FLAG_FLUFFY_BLOCKS
+#define P2P_SUPPORT_FLAG_V2                             0x02
```
If v2 adds only notify_tx messages, I would call this flag `P2P_SUPPORT_FLAG_NOTIFY_TX`

Boog900 replied:
I would say `P2P_SUPPORT_FLAG_TX_RELAY_V2`, the "old" message is called `NOTIFY_NEW_TRANSACTIONS`.

0xFFFC0000 replied:
`P2P_SUPPORT_FLAG_TX_RELAY_V2` seems more reasonable. I have nothing against `P2P_SUPPORT_FLAG_NOTIFY_TX`, but `P2P_SUPPORT_FLAG_NOTIFY_TX` name does not show that we are using totally separate method of tx relay.

I will update it in next push.

0xFFFC0000 replied:
Fixed. Feel free to resolve it.

Boog900 commented on src/cryptonote_protocol/levin_notify.cpp:
```diff
@@ -202,12 +203,58 @@ namespace levin
+
+      if (pad)
+      {
+        size_t bytes = 9 /* header */ + 4 /* 1 + 'txs' */ + tools::get_varint_data(request.txs.size()).size();
+        for(auto tx_hash = request.txs.begin(); tx_hash!=request.txs.end(); ++tx_hash)
+          bytes += tools::get_varint_data(sizeof(*tx_hash)).size() + sizeof(*tx_hash);
```
The serialized message stores all hashes in a single field, so this is incorrect.

```suggestion
        size_t bytes = 9 /* header */ + 4 /* 1 + 'txs' */ + tools::get_varint_data(request.txs.size() * 32).size() + request.txs.size() * 32;
```

Although even that is incorrect, it uses the consensus (COCOA BEANS) varint, not the epee varint and misses the field tag size. This is the same as the `make_tx_message` padding calculation. 

@vtnerd does this slight difference in calculation of the size vs the real size matter for the padding?

vtnerd replied:
This is why I promoted the usage of the technique the SSL PR used - it's more accurate, quicker, and easier to just pad the end of the levin message as the extra trailing data gets dropped. Arguably this feature should be dropped entirely (see below).

Assuming this implementation is kept, it's possibly leaking data, but it's unlikely to matter. This feature is only useful if the connection is encrypted, which currently occurs only over Tor/I2P, both of which are already padded to disrupt traffic shaping analysis. In other words, Tor/I2P already do what we want here.

When/if SSL gets merged, this should be re-thought as SSL provides no automatic padding, which is why I added randomized padding (at the levin "layer") in that patch. It pads a randomized amount up to 8KiB, but perhaps it should just "round up" like Tor/I2P do (if it works for those networks, surely it works for us).

Boog900 commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -934,25 +1063,41 @@ namespace cryptonote
+      }
+      else
+      {
+        m_peer_info_manager.add_received(context.m_connection_id);
```
IMO this should take into account if we actually requested the tx from the peer. Currently the peer can send us an inv, we request it, they send a different tx and to us it looks like they have sent as many as we requested, when in reality they haven't.

Boog900 commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -883,6 +938,80 @@ namespace cryptonote
+  {
+    MLOG_P2P_MESSAGE("Received NOTIFY_TX_POOL_INV (" << arg.txs.size() << " txes)");
+
+    // Create a list to hold transaction hashes missing in our local tx pool.
+    std::vector<crypto::hash> missing_tx_hashes;
+    std::time_t now = std::time(nullptr);
+
+    // Iterate over each advertised transaction hash and check our pool and our requested tracker.
+    for (const auto &tx_hash : arg.txs)
+    {
+      // If we have the tx in our pool, also remove it from request manager and skip.
+      if (m_core.pool_has_tx(tx_hash))
+      {
+        m_request_manager.remove_transaction(tx_hash);
+        continue;
+      }
+
+      bool need_request = m_request_manager.add_transaction(tx_hash, context.m_connection_id, now);
+      m_peer_info_manager.add_announcement(context.m_connection_id, tx_hash);
+      if (need_request) {
+        m_peer_info_manager.add_requested_from_peer(context.m_connection_id);
+        missing_tx_hashes.push_back(tx_hash);
+      }
+    }
+
+    // If there are missing transactions, send a RequestTxPoolTxs message.
+    if (!missing_tx_hashes.empty())
+    {
+      NOTIFY_REQUEST_TX_POOL_TXS::request req;
+      req.txs = std::move(missing_tx_hashes);
+      post_notify<NOTIFY_REQUEST_TX_POOL_TXS>(req, context);
+      MLOG_P2P_MESSAGE("Requested " << req.txs.size() << " missing transactions via RequestTxPoolTxs");
+    }
+    else
+    {
+      MLOG_P2P_MESSAGE("All advertised transactions are already in the pool or were requested recently");
+    }
+
+    return 1;
+  }
```
Currently from what I can see a peer has 30 seconds to bomb our node with as many tx invs as possible and we have to track each one. I think we should a limit on the amount of inflight requests we make to a single peer.

So if we have more than like 50 requests to a single peer we should ignore all future invs from that peer until that number drops.

0xFFFC0000 replied:
Addressed in the new push. Feel free to resolve this issue if you consider the implementation adequate.

Boog900 reviewed (commented):
First pass, thanks for working on this :pray:

vtnerd commented on src/cryptonote_protocol/txrequesthandler.h:
```diff
@@ -0,0 +1,129 @@
+  tx_request_handler &operator=(tx_request_handler &&other) = delete;
+  tx_request_handler &operator=(const tx_request_handler &other) = delete;
+  tx_request_handler *operator&() = delete;
+  bool operator>(const tx_request_handler &other) const = delete;
```
C++ doesn't define defaults for these, just remove them and the compiler will generate errors.

0xFFFC0000 replied:
Yes, I know C++ doesn’t define them, it is left over that I didn’t want to touch. Since don’t want anyone to abuse this class in future.

vtnerd commented on src/cryptonote_protocol/txrequesthandler.h:
```diff
@@ -0,0 +1,129 @@
+
+private:
+  // Thread for checking the transaction requests
+  std::thread m_handler_thread;
```
There has to be a way to implement this without some arbitrary thread tucked away. Admittedly I'm being lazy and not going through everything, but this just has to be the result of poor design choices.

0xFFFC0000 replied:
> This just has to be the result of poor design choices.

That is inaccurate statement. I don’t want to block existing operations. I know already idle gets called once a second. But doing stuff there means blocking it.

vtnerd replied:
At a glance it looks like this `std::thread` "blocks" the ASIO threads with locks. Is that any better? The `std::thread` could be blocked for a while by the ASIO thread (or vice-versa)?

Just use the idle thread timer, it's there specifically for this purpose. As long as the locks are released within a reasonable time-frame it'll be fine. You might be able to pull a strand implementation to prevent all blocking, but that'd be a bit more work I think.

0xFFFC0000 replied:
> You might be able to create a strand implementation to prevent all blocking, but that'd be a bit more work I think. 

This is a better approach IMHO. Let me look up what I can do. 

All of your comments will be addressed in the next update.

vtnerd commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -883,6 +938,80 @@ namespace cryptonote
+  {
+    MLOG_P2P_MESSAGE("Received NOTIFY_REQUEST_TX_POOL_TXS (" << arg.txs.size() << " txes)");
+
+    // Create a response (for new transactions)
+    NOTIFY_NEW_TRANSACTIONS::request response = {};
+
+    // Iterate over requested txin hashes
+    for (const auto &tx_hash : arg.txs)
+    {
+      m_peer_info_manager.add_requested_from_me(context.m_connection_id);
+      // Attempt to get the transaction blob from the mempool;
+      cryptonote::blobdata tx_blob;
+      if (m_core.get_pool_transaction(tx_hash, tx_blob, cryptonote::relay_category::all))
+      {
+        m_peer_info_manager.add_sent(context.m_connection_id);
+        response.txs.push_back(std::move(tx_blob));
+      }
+      // If tx is not in the pool, then ignore it (do not penalize peer)
+    }
+
+    // Send response if any txs found
+    if (!response.txs.empty())
+    {
+      post_notify<NOTIFY_NEW_TRANSACTIONS>(response, context);
+    }
```
* Padding mostly helps with Tor/I2P when noise is OFF. In this PR, only number of hashes is potentially leaked, so I don't see a massive need for including it.
* Padding this way should be avoided in favor of what is done in my [SSL patch](https://github.com/monero-project/monero/pull/8996/files). See `levin_base.h` and `levin_base.cpp`. The inner epee format is self-describing so you can safely write random bytes after the payload, and they will get dropped/ignored. This is quicker, more efficient, and less error prone than doing it with the `_` field.
  * The SSL will deprecate this padding code if merged first, so might as well just import it.

Boog900 replied:
I agree padding is not necessary for the new messages. Although here is where we are sending the full txs in response to a request for them.
 
 The old protocol looked like this:
 
 node --- txs (maybe padded) --> node
 
 the new protocol:
 
 node --- inv  (not  padded) --> node
 node <-- request  (not padded) --- node
 node --- txs  (not padded) --> node
 
  my question was about the last step. IMO we should pad to match the old protocol. Although if we never fluff over Tor/I2p this isn't really necessary.

0xFFFC0000 replied:
Added padding here. Fixed. Feel free to resolve it.

https://github.com/monero-project/monero/compare/3fcd5c70cf964b4f283c92538c81f2f2aed84c53..f345aea15430f7bbc79e6b613baba9a6bd245bf8

vtnerd commented on src/cryptonote_protocol/peerinfo_manager.h:
```diff
@@ -0,0 +1,413 @@
+#include "string_tools.h"
+#include "syncobj.h"
+
+class peer_info_manager {
```
This `class` shouldn't exist - you can fetch by `uuid`/`connection_id` already. Instead, the information should be tucked into the `connection_context` object.

This implementation creates several problems:
  * You have to separately prune the map when connections close
  * Everything EXCEPT "tx inventory" is stored in the `connection_context` object

0xFFFC0000 replied:
I disagree. Connection context is mostly networking / connection related information. I am trying to have peer abstraction here without the networking information. 

Eventually I want to have well defined statistical scoring system that would score the peer based on its performance. If this was more networking related information, your proposal would’ve made more sense. But at this abstraction layer I disagree with your proposal.

0xFFFC0000 replied:
Your criticism about using `connection_context` is reasonable.

0xFFFC0000 replied:
All of your comments will be addressed in the next update.

0xFFFC0000 replied:
Feel free to resolve it.  I moved everything inside `connection_context`, and got rid of this class.

vtnerd commented on src/cryptonote_protocol/peerinfo_manager.h:
```diff
@@ -0,0 +1,413 @@
+    struct peer_info_state {
+      boost::uuids::uuid connection_id;
+      std::unordered_set<crypto::hash> announcements;
+      size_t received = 0;
```
Do we really need to track all of this information? Is it useful for debugging somehow?

0xFFFC0000 replied:
Yes. That is the problem with our existing code. Used haphazard, temporary methods. And trying to fix later. 

Instead I am keeping all the info in central, well defined place. That can be used later. Possibly with a better way to statistically score peers and remove them if the score drops.

vtnerd replied:
> Yes. That is the problem with our existing code. Used haphazard, temporary methods. And trying to fix later.

This is somewhat vague criticism. Do other systems of similar size have permanent (never needing maintenance) code? Zero bugs or something?

> No industry grade professional engineer is thinking about keeping few bytes, that can be used later, as problem.

I never said it would be a memory issue; we sometimes track information of limited use in debugging. These field dumps can just clutter the log further in the worse case. I'm on the fence whether these fields help with debugging. If they helped you at some point, then perhaps they are useful.

I find that that tracing when particular events occur is more useful. Because you don't really know ordering or timing with these kinds of dumps, just that event X occurred Y times.

vtnerd commented on src/cryptonote_protocol/peerinfo_manager.h:
```diff
@@ -0,0 +1,413 @@
+      return m_data->missed;
+    }
+
+    std::string get_info() const {
```
Just make this `std::ostream operator<<(...)`. Its more efficient.

vtnerd replied:
I got a bit lazy and didn't see if it was being used outside of `std::ostream` cases. So safely ignore if not always in use.

vtnerd commented on src/cryptonote_protocol/peerinfo_manager.h:
```diff
@@ -0,0 +1,413 @@
+  void add_peer(const boost::uuids::uuid &id) { m_peer_info.emplace(id); }
+
+  void remove_peer(const boost::uuids::uuid &id) {
+    auto it = std::find_if(m_peer_info.begin(), m_peer_info.end(),
```
You should be using an `unordered_map<id, data>` instead of an `unordered_set` and then all of these `find_if` calls become efficient hash lookups instead of log(n) iteration. Another reason to use the built-in connection tracking, which does this (relatively) efficiently already.

0xFFFC0000 replied:
That is inaccurate statement IMHO. unordered_set and unordered_map are using same underlying dtat structure. For example, IIRC libc++ is using redblack tree to implement both. I think you are mistaken `std::unordered_set`  with `std::set`


Both std::unordered_set and std::unordered_map offer average-case O(1) time for insert, find, and erase (hash-based) operations

vtnerd replied:
And to clarify a bit, the design makes more sense if the `shared_ptr` was being used outside of this class. But the way its being used here doesn't seem helpful.

0xFFFC0000 replied:
All of your comments will be addressed in the next update.

vtnerd replied:
A `shared_ptr` was not useful originally, but may be useful after the ongoing re-design.

0xFFFC0000 commented on src/cryptonote_protocol/cryptonote_protocol_defs.h:
```diff
@@ -375,5 +375,51 @@ namespace cryptonote
+      std::string _; // padding
+
+      BEGIN_KV_SERIALIZE_MAP()
+        KV_SERIALIZE_CONTAINER_POD_AS_BLOB(txs)
```
Fixed. Feel free to resolve it. 

https://github.com/monero-project/monero/compare/f2061bf8976e3872e7d227dd5ee365213a317a4c..d024d369d590a0bd6488757266fee0c45df217c7

nahuhh commented:
Not sure if this is preexisting on master/release:

testing was done on a testnet with 10mb blocks.

during tx relay, sending all txs from Node A keeps a very low "received" amount, as shown via `print_net_stats`.
Node C, which only received txs _and_ mines blocks, is the opposite.

The strange part, to me, is: when Node C produces a block, it sends quite a large block. IIUC, fluffyblocks should mean that Node C should be sending a minimal amount of data, since Node A already has _100%_ of the txs associated with the block.

here's an example of print_net_stats from Node A before and after receiving a block.

<details>
<summary>print_net_stats</summary>

Before
```
print_net_stats
Received 90371 bytes (88.25 kB) in 48 packets in 7.0 minutes, average 214 B/s = 0.00% of the limit of 32.00 MB/s
Sent 7467544 bytes (7.12 MB) in 952 packets in 7.0 minutes, average 17.32 kB/s = 0.21% of the limit of 8.00 MB/s
```

After
```
2025-09-13 23:43:20.634 I Synced 1478/1478
print_net_stats
Received 6581613 bytes (6.28 MB) in 848 packets in 7.3 minutes, average 14.78 kB/s = 0.05% of the limit of 32.00 MB/s
Sent 7675842 bytes (7.32 MB) in 979 packets in 7.3 minutes, average 17.23 kB/s = 0.21% of the limit of 8.00 MB/s
```
</details>

j-berman reviewed (commented):
I haven't had the chance to review this at length yet, but it would be nice if it also implemented some max limit of txs when responding to `NOTIFY_REQUEST_TX_POOL_TXS`, as opposed to relaying all missing txs in a single response. This avoids problems when the pool exceeds the max packet size limit.

EDIT: this suggestion is a weakly held suggestion and not a requirement for this PR, since this PR only applies to brand new txs as they are relayed on the network.

~~Ultimately, I also think this PR could replace the txpool complement fetching that nodes currently do (`request_txpool_complement`).~~ (edit: ignore this, we'd still want the txpool complement in addition to this)

0xFFFC0000 commented on src/cryptonote_protocol/txrequestqueue.h:
```diff
@@ -0,0 +1,87 @@
+struct request {
+  boost::uuids::uuid peer_id;
+  crypto::hash tx_hash;
+  std::time_t firstseen_timestamp;
```
Fixed: https://github.com/monero-project/monero/pull/9933/commits/64121e9df24f60b071971edef4a80d9120ce5a8e

0xFFFC0000 commented on src/cryptonote_protocol/txrequestqueue.h:
```diff
@@ -0,0 +1,87 @@
+#include "common/util.h"
+
+
+struct request {
```
Fixed: https://github.com/monero-project/monero/pull/9933/commits/2875352e33ed4b401a90daff6d8e3a71dbd8cda5

0xFFFC0000 commented on src/cryptonote_protocol/cryptonote_protocol_handler.h:
```diff
@@ -36,6 +36,7 @@
 
 #include <boost/program_options/variables_map.hpp>
 #include <string>
+#include <thread>
```
Fixd: https://github.com/monero-project/monero/pull/9933/commits/652b4189710dcacf461e85c5f6f8f928caeff52a

0xFFFC0000 commented on src/cryptonote_protocol/request_manager.h:
```diff
@@ -0,0 +1,248 @@
+      return false;
+
+    }
+    m_requested_txs.insert(request(id, tx_hash, first_seen));
```
Fixed: https://github.com/monero-project/monero/pull/9933/commits/652b4189710dcacf461e85c5f6f8f928caeff52a

0xFFFC0000 commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -934,25 +1041,42 @@ namespace cryptonote
+        if (cryptonote::parse_and_validate_tx_from_blob(tx_blob, tx, tx_hash)
+            && m_request_manager.already_requested_tx(tx_hash))
+        {
+          m_request_manager.remove_requested_transaction(context.m_connection_id, tx_hash);
```
I need both to do a lookup.

0xFFFC0000 commented:
Not necessarily incorrect,, need investigation:

> failure rate calculation


> In request_from_next_peer, reassigning the index reference *might be* wrong


> Erasing requests while iterating them in `check_tx_request_queue`


> What will happen if multiple peers have requested the same TX? a block confirmation may leave one peer’s in_flight_requests count incorrect.

SChernykh commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -1565,6 +1691,37 @@ namespace cryptonote
+                    if (m_request_manager.get_current_request_peer_id(h, owner) && !owner.is_nil())
+                    {
+                        m_p2p->for_connection(owner, [&](cryptonote_connection_context &ctx, nodetool::peerid_type, uint32_t)->bool {
+                            ctx.remove_in_flight_request();
```
If you call `remove_in_flight_request` here, but that peer already received it and we haven't got the response yet (ping is not 0, so network delays are in play), will it `remove_in_flight_request` again when this response arrives?

nahuhh commented:
This pr is missing https://github.com/seraphis-migration/monero/commit/b144d39b01924b5d6f7fb1c5ba7a61c4541d6f93

navidR committed:
tx relay v2: rework tx request management and relay logic

    Implement new relay logic with threshold-based peer dropping, single-peer
    tx hash requests, per-connection in-flight limits with queuing, and
    multiindex-based request tracking. Consolidate add/remove request paths,
    pass tx hashes directly into send_txs/relay_txs, and add unit tests for
    the request manager.

    Co-authored-by: j-berman <justinberman@protonmail.com>
    Co-authored-by: Boog900 <boog900@tutanota.com>

j-berman committed:
seraphis-migration@b144d39

vtnerd commented on src/cryptonote_basic/connection_context.h:
```diff
@@ -32,14 +32,59 @@
 namespace cryptonote
 {
+  //! \return Maximum number of bytes permissible for `command`.
+  inline std::size_t get_command_max_bytes(const int command) noexcept
```
Why was it moved to the header though? This isn't going to be executed as `constexpr`, as it requires a runtime value, so why move it to the header?

vtnerd commented on src/cryptonote_protocol/cryptonote_protocol_handler.inl:
```diff
@@ -876,33 +907,108 @@ namespace cryptonote
+    if (!txs.empty())
+    {
+      NOTIFY_NEW_TRANSACTIONS::request request = {};
+      request.txs = std::move(txs);
```
See this [patch](https://github.com/seraphis-migration/monero/pull/297) in seraphis branches. You'll need to add something similar here - the message should be sending txes in fluff mode.

navidR committed:
address review #discussion_r2893120646

jeffro256 commented on tests/unit_tests/request_manager.cpp:
```diff
@@ -0,0 +1,172 @@
+
+#include "cryptonote_protocol/request_manager.h"
+
+#include <boost/uuid/uuid.hpp>
```
```suggestion
#include <boost/uuid/uuid.hpp>
#include <unistd.h>
```

Adding header include fixes Msys2 builds

selsta commented on src/cryptonote_protocol/request_manager.cpp:
```diff
@@ -0,0 +1,186 @@
+          << " from peer " << epee::string_tools::pod_to_hex(it->peer_id)
+          << ", age: " << elapsed.count() << "ms");
+
+    // If this peer has missed too many requests, we want to drop it
+    if (this->missed_request(it->peer_id, 1)) {
+      drop_peers_out.insert(it->peer_id);
+    }
```
```suggestion
    const auto &peer_id = it->peer_id;

    if (m_connection_stats[peer_id].in_flight_requests > 0)
      --m_connection_stats[peer_id].in_flight_requests;

    // If this peer has missed too many requests, we want to drop it
    if (this->missed_request(peer_id, 1)) {
      drop_peers_out.insert(peer_id);
    }
```

https://github.com/seraphis-migration/monero/pull/384

j-berman commented on src/cryptonote_protocol/request_manager.cpp:
```diff
@@ -0,0 +1,186 @@
+#include <cstdint>
+
+void request_manager::remove_peer(const boost::uuids::uuid &peer_id) {
+  MINFO("Removing all requests for disconnected peer: " << epee::string_tools::pod_to_hex(peer_id));
```
nit: I think we should replace including `string_tools.h` with `#include <boost/uuid/uuid.hpp>`, and log all `peer_id`'s in this file like this ` << peer_id` without the `pod_to_hex`.

Can see elsewhere in the code this is done, e.g. [here](https://github.com/monero-project/monero/blob/49c9cfc31568e6e56238a7f7bc8814a4f0c59b24/src/cryptonote_protocol/levin_notify.cpp#L568). boost places the UUID -'s when logging, and `pod_to_hex` doesn't (`2fb59d58-6c29-4579-a79e-4d8f1534a957` vs `2fb59d586c294579a79e4d8f1534a957`). So reusing boost's << operator here actually makes it easier to see what's happening in the logs by making it easier to grep for a specific peer id and see what's happening with it in other places.

navidR replied:
Applied, thanks; peer IDs in request_manager.cpp now stream as Boost UUIDs directly, and the unused string_tools.h include was removed.

navidR committed:
Log request manager peer IDs as UUIDs
